### PR TITLE
fix pca_lowrank memory consumption

### LINKED ISSUE #40768: pca_lowrank memory allocation
The code below allocates a large amount of memory (same size as the input A) in order to subtract the mean.
This can probably be avoided...
https://github.com/pytorch/pytorch/blob/b9cca4b186797b4ccdc12197b3b71db94d8e2297/torch/_lowrank.py#L273

cc @VitalyFedyunin @ngimel

### PR DESCRIPTION
**Pull request by ngimel**
Per title, fixes #40768

### CODE REVIEW

**Review by pearu — APPROVED** [review]
LGTM, thanks for this simplification!

File: torch/_lowrank.py [hunk 1]
```diff
@@ -267,8 +267,5 @@ def pca_lowrank(A, q=None, center=True, niter=2):
         M = _utils.transpose(torch.sparse.mm(C_t, ones_m1_t))
         return _svd_lowrank(A, q, niter=niter, M=M)
     else:
-        c = A.sum(dim=(-2,)) / m
-        C = c.reshape(A.shape[:-2] + (1, n))
-        ones_m1 = torch.ones(A.shape[:-1] + (1, ), dtype=dtype, device=A.device)
-        M = ones_m1.matmul(C)
-        return _svd_lowrank(A - M, q, niter=niter, M=None)
+        C = A.mean(dim=(-2,), keepdim=True)
+        return _svd_lowrank(A - C, q, niter=niter, M=None)
```
Comment (pearu) [inline torch/_lowrank.py:271]: Could you change this code to:
```
        C = A.mean(dim=(-2,), keepdim=True)
        M_strides = [C.stride(i) for i in range(len(C.shape))]
        M_strides[-2] = 0
        M = C.as_strided(A.shape, M_strides)
        return _svd_lowrank(A, q, niter=niter, M=M)
```
that will avoid creating a possibly large `A - C`.

**Review by pearu — CHANGES_REQUESTED** [review]
Suggestion to use stride tricks to avoid creating possible large temporary tensors.

File: torch/_lowrank.py (hunk 1 above)
Comment (ngimel) [reply torch/_lowrank.py:271]: Will do that, but I suspect it will just postpone the inevitable - when matmul sees discontiguous arguments, it blow them up. Will have to double check matmul code, but I vaguely remember it from when I looked at it previously.

Comment (ngimel) [reply torch/_lowrank.py:271]: Do you ever backprop through pca_lowrank? If not, and A is otherwise not needed, A-C can be computed inplace, avoiding large memory allocation.

Comment (pearu) [reply torch/_lowrank.py:271]: > Will do that, but I suspect it will just postpone the inevitable - when matmul sees discontiguous arguments, it blow them up

I tested the code locally, all pca tests passed ok.

Comment (ngimel) [reply torch/_lowrank.py:271]: Yeah, sorry, by "blow up" I meant it will materialize expanded contiguous tensors. So yeah, the tests will pass, but memory usage is possibly still suboptimal.

Comment (pearu) [reply torch/_lowrank.py:271]: > Do you ever backprop through pca_lowrank? If not, and A is otherwise not needed, A-C can be computed inplace, avoiding large memory allocation.

The pca_lowrank algorithm is designed for large sparse inputs that avoid creating any large temporary tensors. Changing the inputs in-place would be a regression, imho. The current issue is with dense inputs only that required computing `A-C`, but I believe the stride trick provides the workaround and would actually simplify also the pca_lowrank implementation.

Comment (pearu) [reply torch/_lowrank.py:271]: > I meant it will materialize expanded contiguous tensors.

If that is the case then I think this is a bug in matmul: any strided tensor is already materialized IIUC, the underlying algorithms should not check the stride values but use these as they are.

Comment (pearu) [reply torch/_lowrank.py:271]: My statement holds only when `M` is contiguous in memory. I did not check that, so @ngimel, your concern is valid.
If `M` in the above is not contiguous, then perhaps this is:
```
M = C.contiguous().as_strided(A.shape, M_strides)
```

Comment (pearu) [reply torch/_lowrank.py:271]: I think my original suggestion does not resolve the underlying issue because `M` and `M_H` are both needed for the algorithm and for > 2 D cases one of these is always non-contiguous and this leads to the expansion of memory.
So, the suggestion needs to be modified:
```
        C = A.mean(dim=(-2,), keepdim=True)
        M_strides = [C.stride(i) for i in range(len(C.shape))]
        M_strides[-2] = 0
        M = C.as_strided(A.shape, M_strides)
        M_H = C.transpose(-2, -1).contiguous().as_strided(<shape of A.transpose(-2, -1)>, <M_H_strides>)
        return _svd_lowrank(A, q, niter=niter, M=M, M_H=M_H)
```
The above requires a little research to ensure that no expansions of memory will happen and what is the correct strides to `M_H`.

So, in short term, `return _svd_lowrank(A - C, q, niter=niter, M=None)` is better than my original suggestion.

For best memory and processor efficiency, suggest users to center `A` in-place and use `center=False`.

Comment (pearu) [reply torch/_lowrank.py:271]:
```
>>> a=torch.randn(1,1).expand(10,10)
>>> a.stride()
(0, 0)
>>> a.is_contiguous()
False
```
This looks wrong because `a` memory is contiguous: it contains just one value.

Anyway, the suggestions to prepare `M` and `M_H` using stride tricks will not work because of the above observation.

Comment (ngimel) [reply torch/_lowrank.py:271]: `a` memory is contiguous, but tensor a is not contiguous. In contiguous tensors, no 2 logically different elements point to the same memory location, whereas in a all elements point to the same memory location.

Comment (pearu) [reply torch/_lowrank.py:271]: Right, the whole point of expand-contiguous conversion is to ensure that tensor data is C-contiguous and can be passed to some low-level routine that is not stride-aware, that is, the low-level would access the data memory according to the shape information, not strides. That said, matmul low-level routines ought to be stride-aware (thinking of LAPACK) so that expand-contiguous conversion is unnecessary. So there could be some ideas for matmul optimization here.

**Review by pearu — APPROVED** [review]
Approving the PR as it is because my suggestion was invalid.

File: torch/_lowrank.py (hunk 1 above)
Comment (ngimel) [reply torch/_lowrank.py:271]: I hear what you are saying, however, typically matmul (let's talk about 2d tensors for simplicity) will be dispatched to a blas library that implements [<X>gemm API ](http://ps-2.kev009.com/tl/techlib/manuals/adoclib/libs/basetrf2/sgemm.htm). As you can see here, lda/ldb that are providing stride-awareness are explicitly constrained to be positive, not 0. That said, instead of dispatching calls with stride=0 args to blas, doing pointwise multiplication of arguments followed by a reduction and expanding back to the necessary size is likely to be faster. Double points if multiplication could be done on the fly during reduction, so that A*B (pointwise-multiplied) did not have to be materialized, but unfortunately pytorch cannot do that, you'd need keops for that.

Comment (ezyang) [reply torch/_lowrank.py:271]: surprised that blas doesn't support zero stride :( Do we need to write specialized kernels for these cases?

Comment (ngimel) [reply torch/_lowrank.py:271]: The joys of 80's libraries for you :-) The primitive that we need here to avoid materialization is "broadcasted multiplication followed by a reduction" https://github.com/pytorch/pytorch/issues/32591#issuecomment-578443836, einsum people have been talking about it for a long time. It sounds easy but, given how different broadcasting patterns can be and how generally hard it is to write an efficient reduction for various access patterns, turns out to be pretty hard, there are libraries that are doing it to varying degrees of success (Keops and cutensor that I know of)
